gadget/gadget.go

```diff
@@ -445,16 +445,59 @@ func readInfo(f func(string) ([]byte, error), gadgetYamlFn string, model Model)
 
 // ReadInfo reads the gadget specific metadata from meta/gadget.yaml in the snap
 // root directory.
+// See ReadInfoAndValidate for a variant that does consistency rules
+// validation like Validate.
 func ReadInfo(gadgetSnapRootDir string, model Model) (*Info, error) {
 	gadgetYamlFn := filepath.Join(gadgetSnapRootDir, "meta", "gadget.yaml")
-	return readInfo(ioutil.ReadFile, gadgetYamlFn, model)
+	ginfo, err := readInfo(ioutil.ReadFile, gadgetYamlFn, model)
+	if err != nil {
+		return nil, err
+	}
+	return ginfo, nil
+
+}
+
+// ReadInfoAndValidate reads the gadget specific metadata from
+// meta/gadget.yaml in the snap root directory.
+// It also performs consistency rules validation as Validate does
+// using the given constraints. See ReadInfo for a variant that does not.
+func ReadInfoAndValidate(gadgetSnapRootDir string, model Model, validationConstraints *ValidationConstraints) (*Info, error) {
+	ginfo, err := ReadInfo(gadgetSnapRootDir, model)
+	if err != nil {
+		return nil, err
+	}
+	if err := Validate(ginfo, model, validationConstraints); err != nil {
+		return nil, err
+	}
+	return ginfo, err
 }
 
 // ReadInfoFromSnapFile reads the gadget specific metadata from
 // meta/gadget.yaml in the given snap container.
+// It also performs consistency rules validation as Validate does.
+// See ReadInfoFromSnapFileNoValidate for a variant that does not.
 func ReadInfoFromSnapFile(snapf snap.Container, model Model) (*Info, error) {
+	ginfo, err := ReadInfoFromSnapFileNoValidate(snapf, model)
+	if err != nil {
+		return nil, err
+	}
+	if err := Validate(ginfo, model, nil); err != nil {
+		return nil, err
+	}
+	return ginfo, nil
+}
+
+// ReadInfoFromSnapFileNoValidate reads the gadget specific metadata from
+// meta/gadget.yaml in the given snap container.
+// See ReadInfoFromSnapFile for a variant that does consistency rules
+// validation like Validate as well.
+func ReadInfoFromSnapFileNoValidate(snapf snap.Container, model Model) (*Info, error) {
```

Comment:
thanks for these new functions, while awkward and a bit of a mouthful, I think these will make for more readable code at least that is harder to get wrong when using it